src/sys/socket/mod.rs

```diff
@@ -556,243 +566,141 @@ pub enum ControlMessageOwned {
+impl ControlMessage for ControlMessageOwned {
+    fn decode(header: &cmsghdr, buf: &[u8]) -> Option<Self> {
+        macro_rules! ok {
+            ( $t:ident ) => { ok!($t, $t) };
+            ( $t:ident, $name:ident ) => {
+                if let Some(cmsg) = $t::decode(header, buf) {
+                    return Some(ControlMessageOwned::$name(cmsg));
                 }
-                let cmo = ControlMessageOwned::ScmRights(fds);
-                cmo
-            },
-            #[cfg(any(target_os = "android", target_os = "linux"))]
-            (libc::SOL_SOCKET, libc::SCM_CREDENTIALS) => {
-                let cred: libc::ucred = ptr::read_unaligned(p as *const _);
-                ControlMessageOwned::ScmCredentials(cred)
-            }
-            (libc::SOL_SOCKET, libc::SCM_TIMESTAMP) => {
-                let tv: libc::timeval = ptr::read_unaligned(p as *const _);
-                ControlMessageOwned::ScmTimestamp(TimeVal::from(tv))
-            },
-            #[cfg(any(
-                target_os = "android",
-                target_os = "freebsd",
-                target_os = "ios",
-                target_os = "linux",
-                target_os = "macos"
-            ))]
-            (libc::IPPROTO_IPV6, libc::IPV6_PKTINFO) => {
-                let info = ptr::read_unaligned(p as *const libc::in6_pktinfo);
-                ControlMessageOwned::Ipv6PacketInfo(info)
-            }
-            #[cfg(any(
-                target_os = "android",
-                target_os = "ios",
-                target_os = "linux",
-                target_os = "macos"
-            ))]
-            (libc::IPPROTO_IP, libc::IP_PKTINFO) => {
-                let info = ptr::read_unaligned(p as *const libc::in_pktinfo);
-                ControlMessageOwned::Ipv4PacketInfo(info)
-            }
-            #[cfg(any(
-                target_os = "freebsd",
-                target_os = "ios",
-                target_os = "macos",
-                target_os = "netbsd",
-                target_os = "openbsd",
-            ))]
-            (libc::IPPROTO_IP, libc::IP_RECVIF) => {
-                let dl = ptr::read_unaligned(p as *const libc::sockaddr_dl);
-                ControlMessageOwned::Ipv4RecvIf(dl)
-            },
-            #[cfg(any(
-                target_os = "freebsd",
-                target_os = "ios",
-                target_os = "macos",
-                target_os = "netbsd",
-                target_os = "openbsd",
-            ))]
-            (libc::IPPROTO_IP, libc::IP_RECVDSTADDR) => {
-                let dl = ptr::read_unaligned(p as *const libc::in_addr);
-                ControlMessageOwned::Ipv4RecvDstAddr(dl)
-            },
-            (_, _) => {
-                let sl = slice::from_raw_parts(p, len);
-                let ucmsg = UnknownCmsg(*header, Vec::<u8>::from(&sl[..]));
-                ControlMessageOwned::Unknown(ucmsg)
-            }
+            };
         }
+
+        ok!(ScmRights);
```

Comment:
What's the point of turning ControlMessage into a trait if you still have to hard-code every option here?

Reply:
This allows developers to define their own `ControlMessage` without having to fork nix.

You can define your own enum to avoid space and performance overhead.

Reply:
Can you show an example of decoding a cmsg that is defined outside of Nix?

Reply:
I used `TLS_SET_RECORD_TYPE` and `IP_RECVERR` in ktls and zero-copy tcp, which is not defined in `nix`.

With trait, I can define my own `ControlMessage`,

```rust
struct IpRecvError(IpType, SockExtendedErr);

impl ControlMessage for IpRecvError {
    ...
}

// ...

let IpRecvError(_, err) = cmsg.next::<IpRecvError>()?;

// ...
```

Reply:
But that crate manually casts a raw pointer to downcast the control message to a specific type.  Do you have a plan for a safe API instead?

Reply:
Yes, I want to use nix.

Reply:
But I don't see such an API in your PR.  What I'm asking is, "using Nix, how would a user safely decode a cmsg into a type not implemented within Nix itself?

Reply:
Users can parse bytes. If user uses raw ptr conversion, the user needs to ensure security.

Reply:
If the users want to deal with raw pointers, then they may as well be using libc.  The purpose of Nix is to provide a safe and idiomatic API on top of libc.  We shouldn't be exposing raw pointer APIs except where absolutely necessary.

Reply:
We didn't expose the raw ptr, we exposed a `&[u8]`. For some cmsg, this does not require unsafe.

For example `TLS_GET_RECORD_TYPE`, we can write

```rust
pub struct TlsGetRecordType(u8);

impl ControlMessage for IpRecvError {
    fn decode(header: &cmsghdr, buf: &[u8]) -> Option<Self> {
        if header.cmsg_level == SOL_TLS && header.cmsg_type == TLS_GET_RECORD_TYPE {
            Some(TlsGetRecordType(buf[0]))
        } else {
            None
        }
    }
}
```

If you think that nix maintains a huge cmsg enum is correct, then I have no opinion.

Reply:
What I've been asking for is an **example** of how to use the API from this PR.  I don't want to see a crate that isn't even using Nix, and I don't want an incomplete English-language explanation.  I want an **example**, that includes defining a new `ControlMessage` implementation along with sending and receiving it.

Reply:
I can't use nix because nix doesn't currently support the cmsg type I need. I misunderstood your question, I thought you were asking me where I needed it.

I already said at the beginning that this PR is incomplete, just showing the possibility of a user defining their own cmsg type.

Anyway, let's end this discussion.